File: lib_net/src/main/java/com/tdxtxt/net/observer/AbsObserverNetapi.java
package com.tdxtxt.net.observer;

import com.tdxtxt.net.NetMgr;
import com.tdxtxt.net.config.NetProvider;
import com.tdxtxt.net.model.AbsResponse;
import org.jetbrains.annotations.NotNull;
import io.reactivex.observers.DisposableObserver;

/**
 * <pre>
 *     author : tangdexiang
 *     time   : 2023-09-14
 *     desc   : 这里不用kotlin，是因为没有边界的泛型R导致无法重写onNext方法，kotlin高版本会有这个问题
 * </pre>
 */
public abstract class AbsObserverNetapi<R> extends DisposableObserver<R>{
    public abstract String host();
    /**
     * errorBody:响应报文中的ErrorBody数据内容
     */
    public abstract void onFailure(int errorCode, String errorMsg, String errorBody);

    public abstract void onSuccess(R response);

    @Override
    public void onNext(@NotNull R response) {
        onSuccess(response);

        onComplete();
    }

    @Override
    public void onError(@NotNull Throwable e) {
        NetProvider provider = getProvider();
        if(provider != null){
            String message = provider.throwable2Message(e);
            int code = provider.throwable2Code(e);
            AbsResponse data = provider.throwable2Response(e);
            String errorBody  = provider.throwable2ErrorBody(e);
            provider.handleError(data, code, message);
            onFailure(code, message, errorBody);
        }else{
            onFailure(-999, "provider is null", "");
        }
        onComplete();
    }

    @Override
    public void onComplete() {

    }

    public NetProvider getProvider() {
        return NetMgr.getProvider(host());
    }
}

File: lib_net/src/main/java/com/tdxtxt/net/observer/AbsObserverNetapi.kt
package com.tdxtxt.net.observer

import com.tdxtxt.net.NetMgr
import io.reactivex.observers.DisposableObserver

/**
 * <pre>
 *     author : ton
 *     time   : 2023/2/10
 *     desc   :
 * </pre>
 */
abstract class AbsObserverNetapi <R> : DisposableObserver<R>() {
    abstract fun host(): String

    abstract fun onSuccess(response: R)
    /**
     * errorBody:响应报文中的ErrorBody数据内容
     */
    abstract fun onFailure(errorCode: Int?, errorMsg: String?, errorBody: String?)

    override fun onNext(response: R) {
        onSuccess(response)

        onComplete()
    }

    override fun onError(e: Throwable) {
        val provider = getProvider()
        val message = provider.throwable2Message(e)
        val code = provider.throwable2Code(e)
        val data = provider.throwable2Response(e)
        val errorBody = provider.throwable2ErrorBody(e)
        provider.handleError(data, code, message)
        onFailure(code, message, errorBody)
        onComplete()
    }

    override fun onComplete() {
    }

    fun getProvider() = NetMgr.getProvider(host())
}